frontend/apps/reader/modules/readerrolling.lua

```diff
@@ -1086,4 +1107,212 @@ function ReaderRolling:showEngineProgress(percent)
+        logger.info("Upgrading book to latest DOM version:")
+
+        -- Backup metadata.lua
+        local cur_dom_version = self.ui.doc_settings:readSetting("cre_dom_version") or "unknown"
+        if self.ui.doc_settings.filepath then
+            local backup_filepath = self.ui.doc_settings.filepath .. ".old_dom" .. tostring(cur_dom_version)
+            if not lfs.attributes(backup_filepath) then -- backup does not yet exist
+                os.rename(self.ui.doc_settings.filepath, backup_filepath)
+                logger.info("  previous docsetting file saved as", backup_filepath)
+            end
+        end
+
+        -- Migrate all XPointers
+        applyFuncToXPointersSlots(migrateXPointer)
+        logger.info(T("  xpointers updated: %1 unchanged, %2 modified, %3 not found let as-is",
+            nb_xpointers_found - nb_xpointers_changed, nb_xpointers_changed, nb_xpointers_lost))
+
+        -- Set latest DOM version, to be used at next load
+        local latest_dom_version = self.ui.document:getLatestDomVersion()
+        self.ui.doc_settings:saveSetting("cre_dom_version", latest_dom_version)
+        logger.info("  cre_dom_version updated to", latest_dom_version)
+
+        -- Switch to default block rendering mode if this book has it set to "legacy",
+        -- unless the user had set the global mode to be "legacy".
+        -- (see ReaderTypeset:onReadSettings() for the logic of block_rendering_mode)
+        local g_block_rendering_mode = G_reader_settings:readSetting("copt_block_rendering_mode")
+        if g_block_rendering_mode ~= 0 then -- default is not "legacy"
+            if not g_block_rendering_mode then -- nil means: use default
+                g_block_rendering_mode = 3 -- default in ReaderTypeset:onReadSettings()
+            end
+            -- This setting is actually saved by self.ui.document.configurable
+            local block_rendering_mode = self.ui.document.configurable.block_rendering_mode
+            if block_rendering_mode == 0 then
+                self.ui.document.configurable.block_rendering_mode = g_block_rendering_mode
+                logger.info("  block_rendering_mode switched to", g_block_rendering_mode)
+            end
+        end
+
+        -- No need for "if doc:hasCacheFile() then doc:invalidateCacheFile()", as
+        -- a change in gDOMVersionRequested has crengine trash previous cache file.
+    end
+
+    -- Check all xpointers
+    applyFuncToXPointersSlots(checkAndCount)
+    logger.info(T("%1 xpointers checked: %2 found (%3 changed) - %4 lost",
+                nb_xpointers, nb_xpointers_found, nb_xpointers_changed, nb_xpointers_lost))
+    if nb_xpointers_lost > 0 then
+        logger.warn("Lost xpointers:")
+        for k, v in pairs(lost_xpointer_info) do
+            logger.warn("  ", k, ":", v)
+        end
+    end
+
+    local text = _([[
+This book has been first opened, and handled since, by some old version of the rendering code.
```

Comment:
Perhaps it can be slightly more succinct?

> This book was first opened by an older version of the rendering code.
> 
> Bookmarks and highlights can be upgraded to the latest version of the code.
> 
> %1
> 
> Proceed with this upgrade and reload the book?

Reply:
I find that a bit too succint :) but well, as the author of the verbose variant, I can't say much :)
Any other opinion ?
(out for a bit, will rethink that later)

Reply:
I kinda like it, I'd possibly switch the second sentence to "You won't lose your bookmarks & highlights" or something similar.

The end goal being hiding the technicalities, and just basically saying "we need to reload the book".

Reply:
ok ok :)
I'd like to keep the `, and handled since, by...` - because, with just `was first opened by`, there's some continuity/logic missing. ("OK, I opened it previously, but why this upgrade now?")
Fine with keeping it?

And so:
`Bookmarks and highlights can be upgraded to the latest version of the code.`
should replace the 2nd sencence of the first paragraph, or also the 2nd paragraph when all is good (`All your bookmarks and highlights are valid and will be available after the migration`) ?

What about the 2 possible (but rare in practice I think) "Note that..." ?

Reply:
`This book is currently handled by an older version of the rendering code`?

Reply:
That's quite ok, but it doesn't tell _why_ it is currently handled by... to the curious user :)